packages/ionio/src/Artifact.ts

```diff
@@ -20,6 +20,89 @@ interface Function {
   asm: string[];
 }
 
+type ArtifactMapper = (artifact: Artifact) => Artifact;
+
+/**
+ * compile one of the constructor input
+ * @param name name of the input to compile
+ * @param arg argument value as input
+ */
+function withConstructorInput(name: string, arg: Argument): ArtifactMapper {
+  return (artifact: Artifact): Artifact => {
+    // compile the constructor input
+    const constructorInputToRemove = artifact.constructorInputs.find(
+      p => p.name === name
+    );
+    if (!constructorInputToRemove) {
+      throw new Error(`Constructor input ${name} not found`);
+    }
+    const argEncoded = encodeArgument(arg, constructorInputToRemove.type);
+
+    return {
+      ...artifact,
+      // remove compiled constructor input
+      constructorInputs: artifact.constructorInputs.filter(
+        p => p.name !== name
+      ),
+      functions: artifact.functions.map(
+        replaceASMtoken(name, argEncoded.toString('hex'))
+      ),
+    };
+  };
+}
+
+/**
+ * change the name of one of the constructor input
+ * @param toreplace name of the input name to replace
+ * @param replaceby new name of the input
+ */
+function withCustomConstructorInputName(
+  toreplace: string,
+  replaceby: string
+): ArtifactMapper {
+  return (artifact: Artifact): Artifact => {
+    return {
+      ...artifact,
+      constructorInputs: artifact.constructorInputs.map(p =>
+        p.name === toreplace ? { ...p, name: replaceby } : p
+      ),
+      functions: artifact.functions.map(
+        replaceASMtoken(toreplace, `$${replaceby}`)
+      ),
+    };
+  };
+}
+
+function replaceASMtoken(
+  toreplace: string,
+  by: string
+): (f: Function) => Function {
+  return (f: Function): Function => ({
+    ...f,
+    asm: f.asm.map(token => {
+      if (token.slice(1) === toreplace) {
+        return by;
+      }
+      return token;
+    }),
+  });
+}
+
+function composeMappers(...mappers: ArtifactMapper[]): ArtifactMapper {
+  return (artifact: Artifact): Artifact => {
+    return mappers.reduce((acc, mapper) => mapper(acc), artifact);
+  };
+}
+/**
+ * Apply a set of mappers to an artifact
+ */
+function transformArtifact(
+  artifact: Artifact,
+  ...mappers: ArtifactMapper[]
```

Comment:
I personally thought about much simpler: dev should not waste time matching `name` with the mapper, what about simply pass an array of `Argument` , so we match 1:1 with position index based on `constructorInputs` order.

Then a special function should communicate to replace it with a `$template` string, presrving the `type`

ie.

```typescript
function transformArtifact(artifact: Artifact, constructorInputs: Argument[])
```

We the pass something

```typescript
transformArtifact(json, ['0xabcd', TemplateString('tiero'))
```

Reply:
You call the `transformArtifact` two times for two importTemplate?

Reply:
@louisinger

Reply:
what do u think of `WithConstructorInputRenamingByIndex(0, 'tiero')` ?

Reply:
why we can't infer the index by the position in the array? Just to make sure the signature is almost the same as the constructorInputs of Contract